In [1]:
# Import necessary libraries

from keras.models import Sequential
from keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout
from keras.optimizers import Adam
from keras.datasets import mnist
from keras.utils import to_categorical

# Load the MNIST dataset
(x_train, y_train), (x_test, y_test) = mnist.load_data()

11490434/11490434 [==============================] - 0s 0us/step


In [2]:
# Normalize the pixel values of the images
x_train = x_train.astype('float32') / 255
x_test = x_test.astype('float32') / 255

In [3]:
# Reshape the images to add a channel dimension
x_train = x_train.reshape(x_train.shape + (1,))
x_test = x_test.reshape(x_test.shape + (1,))

In [4]:
# Convert the labels to one-hot encoded vectors
y_train = to_categorical(y_train, num_classes=10)
y_test = to_categorical(y_test, num_classes=10)

In [5]:
# Define the input shape of the images
input_shape = (28, 28, 1)

In [6]:
# Create a sequential model
model = Sequential()

In [7]:
# Add convolutional layers with max pooling and dropout
model.add(Conv2D(32, kernel_size=(3, 3), activation='relu', input_shape=input_shape))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Dropout(0.25))

model.add(Conv2D(64, kernel_size=(3, 3), activation='relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Dropout(0.25))

In [8]:
# Flatten the output from convolutional layers
model.add(Flatten())

In [9]:
# Add a fully connected layer with dropout
model.add(Dense(128, activation='relu'))
model.add(Dropout(0.5))

In [10]:
# Add an output layer with 10 units and softmax activation
model.add(Dense(10, activation='softmax'))

In [11]:
# Compile the model with categorical crossentropy loss and Adam optimizer
model.compile(loss='categorical_crossentropy',
              optimizer=Adam(lr=0.001),
              metrics=['accuracy'])

/usr/local/lib/python3.10/dist-packages/keras/optimizers/legacy/adam.py:117: UserWarning: The `lr` argument is deprecated, use `learning_rate` instead.
  super().__init__(name, **kwargs)


In [12]:
# Train the model on the MNIST dataset
history = model.fit(x_train, y_train,
                    batch_size=128,
                    epochs=20,
                    validation_data=(x_test, y_test))

Epoch 1/20
469/469 [==============================] - 12s 6ms/step - loss: 0.3695 - accuracy: 0.8842 - val_loss: 0.0695 - val_accuracy: 0.9784
Epoch 2/20
469/469 [==============================] - 3s 6ms/step - loss: 0.1211 - accuracy: 0.9641 - val_loss: 0.0445 - val_accuracy: 0.9846
Epoch 3/20
469/469 [==============================] - 2s 5ms/step - loss: 0.0927 - accuracy: 0.9722 - val_loss: 0.0359 - val_accuracy: 0.9877
Epoch 4/20
469/469 [==============================] - 2s 5ms/step - loss: 0.0761 - accuracy: 0.9780 - val_loss: 0.0318 - val_accuracy: 0.9895
Epoch 5/20
469/469 [==============================] - 2s 5ms/step - loss: 0.0666 - accuracy: 0.9792 - val_loss: 0.0272 - val_accuracy: 0.9912
Epoch 6/20
469/469 [==============================] - 2s 5ms/step - loss: 0.0600 - accuracy: 0.9819 - val_loss: 0.0247 - val_accuracy: 0.9916
Epoch 7/20
469/469 [==============================] - 3s 6ms/step - loss: 0.0539 - accuracy: 0.9832 - val_loss: 0.0254 - val_accuracy: 0.9908
Epoch

In [13]:
# Evaluate the model on the testing set
score = model.evaluate(x_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Test loss: 0.023016266524791718
Test accuracy: 0.9930999875068665




# **Conclusion:**

---


In this project, we built a machine learning model to recognize handwritten digits using the MNIST dataset. We preprocessed the data by normalizing and reshaping it, and used a convolutional neural network architecture to train the model. We achieved a high test accuracy of **99.31%**.

Overall, this project demonstrates the power of machine learning in image recognition tasks. With further tuning and optimization, this model could potentially be used in various applications, such as digit recognition in check processing or postal mail sorting.


---




